### ISSUE TITLE
Prediction errors on using ninja

### ISSUE THREAD

**Issue by Robyt3** [teeworlds/teeworlds#2461]
A lot of prediction errors are popping up in the client when using ninja on vanilla servers.

When activating ninja from this setup:

![screenshot_2020-02-08_12-28-34](https://user-images.githubusercontent.com/23437060/74084400-fece1c80-4a6e-11ea-9edb-d14c61ab9079.png)

```
prediction error
  1 1940 1886 (1940 1886)
  2 1076 987 (1073 985)
  4 640 384 (512 256)
  11 1940 1886 (1940 1886)
  12 1073 985 (1071 984)
```

When ninja attack is constrained in some directions:

![screenshot_2020-02-08_12-28-36](https://user-images.githubusercontent.com/23437060/74084430-5ff5f000-4a6f-11ea-90df-774e0fd29141.png)

```
prediction error
  1 1895 1912 (1895 1912)
  11 1895 1912 (1895 1912)
```

When contrained in all directions, i.e. standing and attacking in a corner, produces no prediction error.

Not sure how much can and should be done about that, as it currently seems to work enough, but having prediction errors with vanilla gamecore seems odd to me.

**Comment by Dune-jr** [comment]
Yeah, afair, when you ninja, your tee teleports because there is no prediction

**Comment by nheir** [comment]
exactly

https://github.com/teeworlds/teeworlds/blob/ffe839d8b61b42286f478058dffb8da4bf119024/src/game/server/entities/character.cpp#L154-L155

**Comment by axblk** [comment]
There is no prediction for weapon effects (ninja, rocket jumps) at all since they are not part of the gamecore. In 0.6 there was "antiping" which basically imitated the whole gameworld (including weapons/projectiles and other players) instead of just the gamecore on client side. As weapon input is handled differently than movement input, it's not as reliable as gamecore prediction, but especially with high ping it can still make a huge difference. Some players even refuse to upgrade to 0.7 because there is no antiping yet.
While this is a nice addition for custom clients, adding this to the official game is problematic. Because you don't know the input from the other players it's rather an assumption than a prediction. It would also require a lot more shared code between server and client and modding weapons might also cause issues then.

**Comment by Dune-jr** [comment]
Yeah, a better name for this feature would be greedy prediction. Like @axblk said, if the server changes something on that side, or if other players interfere, it will look bad.
I imagine this would work well in game modes where other players do not interfere.

**Comment by AssassinTee** [comment]
Rocketjumps can be kind of predictable, when a grenade has a long enough traveltime so you can predict the explosion

**Comment by Dune-jr** [comment]
> Rocketjumps can be kind of predictable, when a grenade has a long enough traveltime so you can predict the explosion

Unless another player intercepts it

**Comment by axblk** [comment]
Problem is that most rocket jumps happen before the client even knows about the projectile. In order to properly predict all rocket jumps, you need to predict the actual firing of the grenades. This is the harder part of antiping.

**Comment by tempral** [comment]
Mentioning another weird behavior of ninja: 
if the tee has an high speed in a direction and +fire the ninja attack in the opposite direction
it "magically" invert all the speed ! like an elastic spring...
